server/lib/resolverMaker.js

```diff
@@ -4,50 +4,52 @@ Handlebars.registerHelper('toJSON', function (json) {
   return new Handlebars.SafeString(JSON.stringify(json))
 })
 
-module.exports = function (dataSourceTypes, dataSourceClients, resolverMappingsJson) {
+module.exports = function (dataSources, resolverMappings) {
   const resolvers = {
     Query: {},
     Mutation: {},
     Subscription: {}
   }
 
-  _.forEach(resolverMappingsJson, (resolverMapping, resolverMappingName) => {
-    // only setup mappings for queries and mutations at this time
-    if (['Query', 'Mutation'].indexOf(resolverMappingName) < 0) {
-      return
+  _.forEach(resolverMappings, (resolverMapping, resolverMappingName) => {
+    if (_.isEmpty(resolverMapping.type)) {
+      throw new Error('Missing query type for mapping: ' + resolverMappingName)
     }
 
-    _.forEach(resolverMapping, (value, key) => {
-      if (_.isEmpty(value.dataSource)) {
-        throw new Error('Missing data source for mapping: ' + key)
-      }
+    if (_.isEmpty(resolverMapping.dataSource)) {
+      throw new Error('Missing data source for mapping: ' + resolverMappingName)
+    }
 
-      if (_.isEmpty(value.requestMapping)) {
-        throw new Error('Missing request mapping for mapping: ' + key)
-      }
+    if (_.isEmpty(resolverMapping.requestMapping)) {
+      throw new Error('Missing request mapping for mapping: ' + resolverMappingName)
+    }
 
-      if (_.isEmpty(value.responseMapping)) {
-        throw new Error('Missing response mapping for mapping: ' + key)
-      }
+    if (_.isEmpty(resolverMapping.responseMapping)) {
+      throw new Error('Missing response mapping for mapping: ' + resolverMappingName)
+    }
 
-      if (!(value.dataSource in dataSourceTypes)) {
-        throw new Error('Unknown data source "' + value.dataSource + '" for mapping ' + key)
-      }
+    if (!(resolverMapping.dataSource in dataSources)) {
+      throw new Error('Unknown data source "' + resolverMapping.dataSource + '" for mapping ' + resolverMappingName)
+    }
 
-      if (dataSourceTypes[value.dataSource] === 'postgres') {
-        const dataSourceClient = dataSourceClients[value.dataSource]
-        try {
-          resolvers[resolverMappingName][key] = buildPostgresResolver(
-            dataSourceClient,
-            value.requestMapping,
-            value.responseMapping
-          )
-        } catch (ex) {
-          console.log('Error while building Postgres resolver for mapping: ' + key)
-          throw new Error('Error while building Postgres resolver for mapping: ' + key)
-        }
+    if (['Query', 'Mutation'].indexOf(resolverMapping.type) < 0) {
```

Comment:
this conditional can be removed now.

Reply:
Hmm I tried to remove it but some tests failed as a result. They failed because the expected behaviour is to exit early if an Unknown query type is used e.g. if a user specifies a resolverMapping with `type: 'foo'` for example.

Reply:
I think it might be best to leave this in for now but also include `Subscriptions`

Reply:
I don't think we an put any restriction here as resolvers can be attached to any custom type as defined by the user in the GraphQL Schema.

See this example where a bunch of different resolvers are needed for parts of the schema besides Query, Mutation & Subscription
https://github.com/apollographql/starwars-server/blob/master/data/swapiSchema.js#L442-L542

Reply:
Another example.
For 1:M relationships defined in you GraphQL Schema, you'll need to attach a Resolver to the 1 side of that relationship e.g.

```graphql
type Post {
  text: String
  comments: [Comment]
}

type Comment {
  text: String
}
```

A resolver is needed on the `comments` field of the Post type e.g. 
```json
{
  "Post": {
    "comments": function() {/*...*/}
  }
}
```

Reply:
During my work, I added test cases for the current state of the code, as _characterization tests_. I also added handling code for some base cases. I wasn't sure about everything in those base cases, but it is super easy to remove those anyway.

BTW, I like this: improving stuff one step at a time.

Reply:
@david-martin thanks for this great explanation! I will need to make some modifications to one or two tests in order for this to work.